examples/LinesWithPassingTones/src/Model.elm

```diff
@@ -0,0 +1,85 @@
+module Model exposing (Model, Msg(..), init, update)
+
+import MusicTheory.Pitch exposing (Octave, Pitch)
+import MusicTheory.PitchClass exposing (Accidental(..), Letter(..))
+import MusicTheory.Scale
+    exposing
+        ( Pattern
+        , Scale
+        , degree
+        , downTo
+        , loweredDegree
+        , pattern
+        , raisedDegree
+        , startAt
+        , upTo
+        )
+
+
+type alias Model =
+    { line : List Pitch }
+
+
+type Msg
+    = NoOp
+
+
+
+{- Ideally, the signature of Scale.generateLine is:
+
+   generateLine : Scale -> Octave -> Pattern -> List Pitch
+
+   and the call to it below in myLineWithPassingTones returns:
+
+       [ Pitch (PitchClass G Natural) 4
+       , Pitch (PitchClass G Sharp) 4
+       , Pitch (PitchClass A Natural) 4
+       , Pitch (PitchClass C Natural) 5
+       , Pitch (PitchClass A Natural) 4
+       , Pitch (PitchClass A Flat) 4
+       , Pitch (PitchClass G Natural) 4
+       , Pitch (PitchClass E Natural) 4
+       ]
+
+   Where this list of pitches is correctly spelled in terms of the scale of
+   C major, uses sharps for raised degrees and flats for lowered degrees,
+   begins on the correct scale degree and octave, and moves in the directions
+   specified by the Pattern.
+-}
+
+
+myLineWithPassingTones : List Pitch
+myLineWithPassingTones =
+    let
+        startingOctave =
+            octave 4
+
+        scale =
+            Scale.major <| C Natural
+    in
+    Scale.generateLine scale startingOctave myPattern
+
+
+myPattern : Pattern
+myPattern =
+    pattern
+        [ startAt (degree 5)
```

Comment:
Passing `Int` can result in invalid state. What if the scale doesn't have a 6th degree?

What if the 6th degree of the scale is already a lowered pitch class?

`startAt` can be anywhere in the list. I don't like that.

Isn't the direction implicit, why does it have to be specified explicitly here?

Also the accidental is influenced by the next note as opposed to the preceding one, I guess.

E.g. G - Eb is going down. So it's rather Eb and not D#?

But in G - D# - E the D# leads to E, so it makes sense to prefer that over Eb, I would think.

Reply:
> Passing `Int` can result in invalid state. What if the scale doesn't have a 6th degree?

It's a good point. I had intended for scale degrees to "wrap around" modulo the number of degrees, so that you could say 9 and get the equivalent of 2 (the same way jazz chord extensions work). But this would complicate things in the case where the number was less than 1. What would the 0th degree of a scale be?

I think it's clear that we'll need some way to traverse the scale, so integers seemed like a natural way to do that. But it might be better for scales to have a named record of degrees and helper functions for moving up and down.

> What if the 6th degree of the scale is already a lowered pitch class and you specify `loweredDegree 6`?

I had just meant for `loweredDegree` to mean the note a minor 2nd below, not a flat. If the 6th degree of C minor is Ab, the `loweredDegree 6` would be Abb or G.

As we discussed before, though, I think you're right that generating lines this way belongs in client code or a separate package.

> `startAt` can be anywhere in the list. I don't like that.

I think it could be helpful in many cases to generate lines starting from specific scale degrees. Many jazz improvisers think this way. On the other hand, the Luigi use case would be too constrained by this probably, and would want to be able to start a pattern on an arbitrary degree. I discussed that in the description here.

> Isn't the direction implicit, why does it have to be specified explicitly here?

If a pattern went from `degree 6` to `degree 1`, how would the program know whether to jump from A4 to C5 or A4 to C4?

This exposes another flaw in this interface, though. If I wanted to jump from A4 to C6, how would I specify that?

> Also the accidental is influenced by the next note as opposed to the preceding one, I guess.
> 
> E.g. G - Eb is going down. So it's rather Eb and not D#?
> 
> But in G - D# - E the D# leads to E, so it makes sense to prefer that over Eb, I would think.

I disagree. Since any note has multiple enharmonic spellings, semitone distance and direction do not provide enough information for an algorithm to decide which note name you mean. The user or client code should be specifying the interval they want to move by. A major third down from G would be Eb. A diminished fourth down from G would be D#.

I think if we place too much trust in an algorithm to infer what enharmonic spelling the user wanted, we will quickly find cases where it cannot correctly choose between different alternatives.